### ISSUE TITLE
refactoring metricdefinitions logic

### ISSUE THREAD

**Issue by Dieterbe** [raintank/raintank-metric#104]
metric-tank spends way too much time in EnsureIndex and related code.
i see 45~50% on dev and prod during normal operations (not right around the time of chunk boundaries, that's another story)

you can see for yourself, from metric_tank dir:
```
ssh -L 18763:localhost:18763 metric-tank-1.raintank.io  &
go tool pprof http://localhost:18763/debug/pprof/profile
(pprof) top50 -cum
5.44s of 16.45s total (33.07%)
Dropped 380 nodes (cum <= 0.08s)
Showing top 50 nodes out of 204 (cum >= 0.77s)
      flat  flat%   sum%        cum   cum%
         0     0%     0%     14.92s 90.70%  runtime.goexit
     0.02s  0.12%  0.12%      7.61s 46.26%  main.(*Handler).HandleMessage.func1
     0.01s 0.061%  0.18%      7.55s 45.90%  github.com/raintank/raintank-metric/metricdef.EnsureIndex
     0.03s  0.18%  0.36%      6.40s 38.91%  github.com/raintank/raintank-metric/metricdef.GetMetricDefinition
         0     0%  0.36%      3.91s 23.77%  github.com/nsqio/go-nsq.(*Consumer).handlerLoop
     0.02s  0.12%  0.49%      3.91s 23.77%  main.(*Handler).HandleMessage
     0.01s 0.061%  0.55%      3.57s 21.70%  github.com/raintank/raintank-metric/schema.MetricDefinitionFromJSON
     0.01s 0.061%  0.61%      3.48s 21.16%  encoding/json.Unmarshal
         0     0%  0.61%      2.85s 17.33%  encoding/json.(*decodeState).unmarshal
     0.14s  0.85%  1.46%      2.83s 17.20%  runtime.systemstack
     0.02s  0.12%  1.58%      2.78s 16.90%  encoding/json.(*decodeState).value
     0.22s  1.34%  2.92%      2.77s 16.84%  encoding/json.(*decodeState).object
     0.04s  0.24%  3.16%      2.74s 16.66%  fmt.Sprintf
     0.03s  0.18%  3.34%      2.69s 16.35%  gopkg.in/redis%2ev2.(*Client).Get
         0     0%  3.34%      2.63s 15.99%  gopkg.in/redis%2ev2.(*baseClient).Process
     0.02s  0.12%  3.47%      2.63s 15.99%  gopkg.in/redis%2ev2.(*baseClient).run
     0.09s  0.55%  4.01%      2.61s 15.87%  main.(*AggMetric).Add
     0.45s  2.74%  6.75%      2.29s 13.92%  fmt.(*pp).doPrintf
```

the code also has some outdated assumptions and approaches:
1)  every single call to EnsureIndex() for every single incoming metric causes a redis lookup, followed by an ES lookup if not in redis, just to see if we should (re-)index a definition.
2) metricdefs are json, could be msgp
3) we call MetricData.Id() a bit too often, which is pretty expensive (this alone is about 12% of the cpu). Id could just be included in the data.

metric-tank also has its own metaCache to have access to metadata (currently just the interval)

here's what i suggest:

* MT gets a cache for all metricdefs, replaces/extends metaCache
* any incoming metricdef gets added to the metricdefcache in RAM, and every 10min, every metric is updated in ES.
* for a series get request which has a metricdefcache miss, we can guarantee we don't have any data for that series. a fallback to redis/ES may result in an old metricdef that we haven't seen yet, but it wouldn't be helpful. so i suggest not to fallback to querying redis/ES. this matches the notion that you shouldn't query an MT instance if it only just came up.
* if our MT instance only just launched, gently add new defs to redis/ES (without checking if entries already exist) following up the warmup pattern.
* basically redis/ES is never queried by MT anymore.  ES is only used to receive queries from graphite-raintank and graphite_watcher.
* we can probably do away with redis, and our plans for redis-cluster.
* at some point, switch to including the Id as data, and use msgp for metricdefs.

@woodsaj and @ctdk thoughts?

**Comment by woodsaj** [comment]
This definitely needs refactoring.

Things i would like to see.

* set the metricdef.Id on ingestion. (we also need to do that for events to prevent duplicates caused by re-queues)
* get rid of redis.
* periodically, or when first seen, flush all metricdefs to persistent storage. this could be local disk or even Cassandra.
* include a "created" timestamp for each metricdef
* at startup, load the full cache of metricdefs from the persistent storage.
* only update ES when there is a new metricdef or change in existing metricdef.

**Comment by woodsaj** [comment]
created https://github.com/raintank/raintank-metric/pull/107  first commit adds the "Id" field to the metricData schema.   I love that adding fields to the struct just works with msgpack and there is no need to handle versioning. :)

**Comment by Dieterbe** [comment]
@woodsaj why local disk or cassandra to store metricdefs? I think ES can take us far here, especially now that we can remove almost all of the load. I'm thinking at startup, do an ES search with a scroll, which should be fairly efficient (though we'll have to see about that), and only push updates for new/changed metricdefs. that would be the only MT~ES interaction.  I rather not introduce yet another store unless we really need it. local disk doesn't seem like a good idea anyway since we'll be bringing up new MT instances and then syncing the data from one disk to another becomes the problem.

**Comment by ctdk** [comment]
I'd certainly love to get rid of redis.

**Comment by Dieterbe** [comment]
@woodsaj I pushed some updates to the PR. note that per my comment above we're just using ES only for metricdefs. i think there's pretty much one thing left:  set id's on ingestion for metrics and events (in collector controller i presume) can you do this, cause i'm not quite sure how to best do this.

**Comment by woodsaj** [comment]
@Dieterbe will do.  But the changes to metric-tank aren't dependent on the changes being made to the ingestion. If the ingestion sets the id, then it willl be used, else it will just be generated when it is requested.

**Comment by Dieterbe** [comment]
ok

**Comment by Dieterbe** [comment]
looking a lot better now.
the 8core machines went from cpu idle 400 & 500 to 640 & 730: that's more than 100% improvement :)
https://grafana-monitor.raintank.io/dashboard/snapshot/pKRxsRCu2aajtCRHYwF5cM5kfyzEPid5
also all requests served are about 2x faster: https://grafana-monitor.raintank.io/dashboard/snapshot/HqYP6hTOQmgKrsZN0CZRIANMkDo6HwTM
cpu profile
```
~/g/s/g/r/r/metric_tank ❯❯❯ go tool pprof http://localhost:18763/debug/pprof/profile
Fetching profile from http://localhost:18763/debug/pprof/profile
Please wait... (30s)
Saved profile in /home/dieter/pprof/pprof.localhost:18763.samples.cpu.006.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) top30 -cum
6.39s of 26.78s total (23.86%)
Dropped 367 nodes (cum <= 0.13s)
Showing top 30 nodes out of 176 (cum >= 2.18s)
      flat  flat%   sum%        cum   cum%
         0     0%     0%     22.83s 85.25%  runtime.goexit
         0     0%     0%      9.37s 34.99%  github.com/nsqio/go-nsq.(*Consumer).handlerLoop
     0.10s  0.37%  0.37%      9.35s 34.91%  main.(*Handler).HandleMessage
     2.96s 11.05% 11.43%      6.58s 24.57%  runtime.scanobject
         0     0% 11.43%      6.21s 23.19%  runtime.gcBgMarkWorker
     0.08s   0.3% 11.73%      4.95s 18.48%  runtime.gcDrainUntilPreempt
     0.11s  0.41% 12.14%      4.88s 18.22%  fmt.Sprintf
     0.12s  0.45% 12.58%      4.76s 17.77%  runtime.systemstack
     0.13s  0.49% 13.07%      4.63s 17.29%  github.com/raintank/raintank-metric/schema.(*MetricData).GetId
     0.63s  2.35% 15.42%      3.99s 14.90%  fmt.(*pp).doPrintf
     0.23s  0.86% 16.28%      3.32s 12.40%  fmt.(*pp).printArg
     0.04s  0.15% 16.43%      3.19s 11.91%  runtime.mcall
     0.20s  0.75% 17.18%      3.09s 11.54%  main.(*AggMetric).Add
     0.07s  0.26% 17.44%      3.01s 11.24%  runtime.schedule
     0.05s  0.19% 17.63%      2.87s 10.72%  main.(*DefCache).Add
     1.22s  4.56% 22.18%      2.77s 10.34%  runtime.mallocgc
     0.03s  0.11% 22.29%      2.76s 10.31%  github.com/grafana/grafana/pkg/log.Debug
     0.01s 0.037% 22.33%      2.73s 10.19%  github.com/grafana/grafana/pkg/log.(*Logger).Debug
         0     0% 22.33%      2.58s  9.63%  net/http.(*conn).serve
         0     0% 22.33%      2.48s  9.26%  main.Get
         0     0% 22.33%      2.48s  9.26%  main.get.func1
         0     0% 22.33%      2.48s  9.26%  net/http.(*ServeMux).ServeHTTP
         0     0% 22.33%      2.48s  9.26%  net/http.HandlerFunc.ServeHTTP
         0     0% 22.33%      2.48s  9.26%  net/http.serverHandler.ServeHTTP
         0     0% 22.33%      2.36s  8.81%  runtime.bgsweep
         0     0% 22.33%      2.36s  8.81%  runtime.gosweepone
     0.34s  1.27% 23.60%      2.23s  8.33%  runtime.findrunnable
     0.03s  0.11% 23.71%      2.21s  8.25%  runtime.sweepone
         0     0% 23.71%      2.19s  8.18%  runtime.gosweepone.func1
     0.04s  0.15% 23.86%      2.18s  8.14%  runtime.mSpan_Sweep
(pprof) 
```

**Comment by Dieterbe** [comment]
AJ just deployed the id setting to the coll-cons and this is looking much better again.
https://grafana-monitor.raintank.io/dashboard/snapshot/2QIPpaDBMZqX3zbaCHGtQxO6UZ2taAfD

```
~/g/s/g/r/r/metric_tank ❯❯❯ go tool pprof metric_tank http://localhost:18763/debug/pprof/profile
Fetching profile from http://localhost:18763/debug/pprof/profile
Please wait... (30s)
Saved profile in /home/dieter/pprof/pprof.metric_tank.localhost:18763.samples.cpu.003.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) top30 -cum
3.26s of 11.81s total (27.60%)
Dropped 249 nodes (cum <= 0.06s)
Showing top 30 nodes out of 199 (cum >= 0.68s)
      flat  flat%   sum%        cum   cum%
     0.08s  0.68%  0.68%      9.53s 80.69%  runtime.aeshashbody
         0     0%  0.68%      4.62s 39.12%  github.com/nsqio/go-nsq.(*Consumer).redistributeRDY
     0.05s  0.42%  1.10%      4.58s 38.78%  main.(*Handler).HandleMessage
     0.35s  2.96%  4.06%      3.10s 26.25%  main.(*AggMetric).Add
         0     0%  4.06%      2.78s 23.54%  net/http.(*ServeMux).Handle
     0.02s  0.17%  4.23%      2.76s 23.37%  github.com/grafana/grafana/pkg/log.(*FileLogWriter).DoRotate
         0     0%  4.23%      2.75s 23.29%  net/http.(*conn).serve
         0     0%  4.23%      2.74s 23.20%  github.com/grafana/grafana/pkg/log.(*Logger).StartLogger
         0     0%  4.23%      2.69s 22.78%  net/http.(*ServeMux).match
     0.02s  0.17%  4.40%      2.68s 22.69%  main.Get
         0     0%  4.40%      2.68s 22.69%  main.get.func1
     0.04s  0.34%  4.74%      2.44s 20.66%  fmt.(*buffer).WriteRune
     0.17s  1.44%  6.18%      2.01s 17.02%  fmt.(*pp).doPrintf
     0.08s  0.68%  6.86%      1.76s 14.90%  type..eq.struct { lock runtime.mutex; lockOwner *runtime.g; enabled bool; shutdown bool; headerWritten bool; footerWritten bool; shutdownSema uint32; seqStart uint64; ticksStart int64; ticksEnd int64; timeStart int64; timeEnd int64; reading *runtime.traceBuf; empty *runtime.traceBuf; fullHead *runtime.traceBuf; fullTail *runtime.traceBuf; reader *runtime.g; stackTab runtime.traceStackTable; bufLock runtime.mutex; buf *runtime.traceBuf }
     0.07s  0.59%  7.45%      1.64s 13.89%  fmt.(*pp).handleMethods
     0.02s  0.17%  7.62%      1.62s 13.72%  runtime.schedule
     0.04s  0.34%  7.96%      1.57s 13.29%  runtime.injectglist
     1.39s 11.77% 19.73%      1.39s 11.77%  runtime.duffcopy
     0.02s  0.17% 19.90%      1.38s 11.69%  type..hash.[15]runtime.dbgVar
     0.01s 0.085% 19.98%      1.34s 11.35%  main.getTarget
     0.57s  4.83% 24.81%      1.34s 11.35%  runtime.mallocgc
     0.02s  0.17% 24.98%      1.30s 11.01%  fmt.(*pp).catchPanic
     0.11s  0.93% 25.91%      1.29s 10.92%  runtime.findrunnable
     0.02s  0.17% 26.08%      1.07s  9.06%  main.getSeries
         0     0% 26.08%      0.84s  7.11%  main.(*Chunk).String
     0.14s  1.19% 27.27%      0.78s  6.60%  runtime.concatstrings
     0.03s  0.25% 27.52%      0.77s  6.52%  runtime.largeAlloc
         0     0% 27.52%      0.76s  6.44%  main.initMetrics.func1
         0     0% 27.52%      0.71s  6.01%  github.com/raintank/raintank-metric/msg.Format.String
     0.01s 0.085% 27.60%      0.68s  5.76%  main.(*Chunk).Push
(pprof) list DefCache.*Add
Total: 11.81s
ROUTINE ======================== main.(*DefCache).Add in /home/dieter/go/src/github.com/raintank/raintank-metric/metric_tank/defcache.go
      90ms      420ms (flat, cum)  3.56% of Total
         .          .     64:		add(met)
         .          .     65:	}
         .          .     66:	log.Debug("backfilled %d metric definitions", total)
         .          .     67:}
         .          .     68:
      10ms       10ms     69:func (dc *DefCache) Add(metric *schema.MetricData) {
         .       50ms     70:	id := metric.GetId()
         .       10ms     71:	dc.Lock()
         .      260ms     72:	mdef, ok := dc.defs[id]
         .       10ms     73:	dc.Unlock()
         .          .     74:	if ok {
         .          .     75:		//If the time diff between this datapoint and the lastUpdate
         .          .     76:		// time of the metricDef is grater then 6hours, update the metricDef.
      80ms       80ms     77:		if mdef.LastUpdate < metric.Time-21600 {
         .          .     78:			mdef.LastUpdate = metric.Time
         .          .     79:			dc.addToES(mdef)
         .          .     80:		}
         .          .     81:	} else {
         .          .     82:		mdef = schema.MetricDefinitionFromMetricData(metric)
(pprof) 
```

**Comment by Dieterbe** [comment]
the impact on our ES machines is also quite nice.
https://grafana-monitor.raintank.io/dashboard/snapshot/P7SL7YfZMwpy74ZRpFJ9TvtYcilKuXIU
let the record state that these are 4 core machines.
